doc/nyan/engine_types.md

```diff
@@ -0,0 +1,191 @@
+
+	allowedResource : set(Resource) # Allowed ressources?
+	maxAmount : int # maximum amount  that can be carried
+	speed : float # gathered units per second
+
+	# WIP2: proposal: allow different maximum amounts for different resources:
+
+#	allowedResource : orderedset(Resource)
+#	maxAmount : orderedSet(int) # does <10,5,0,10> work? it is a set after all.
+
+	# WIP3: if WIP2 is not possible: implement an 'Ability' that allows to gather only one resource.
+	# A worker would then need 4 gather abilities added to it's 'abilities' set.
+
+CarryAbility(Ability): # WIP: find a better name?
+	# Used by priests for relics and transport ship (both indirectly [animation needed]) and by castles etc. for many types of units.
+	# WIP1: how should this look like? keep it general?
+	# WIP3: HUDElement for releasing every contained unit and setting a meeting point needed. (But not for priests. -> Check for MovableUnit in allowedUnits?
+	#       or use the alternative below for everything but relics.)
+
+	allowedUnits : set(UnitTypes) # what can be collected?
+	maxUnits : int
+	maxUnitsize : int # maximum size of each unit. If we want this, every movable unit should have a size member declared by a base class.
+	tooDamagedRate : float # percentage of health at which no units can be contained any longer.
+
+# Alternative:
+Shelter(Ability):
+	# A 'Unit' which might hold other units. Used by Castle, Towncenter, TransporterShip, but not priest.
+
+	allowedUnits : set(UnitTypes) # what can be collected?
+	maxUnits : int
+	maxUnitsize : int # maximum size of each unit. If we want this, every movable unit should have a size member declared by a base class.
+	collapsingRate : float # percentage of health at which no units can be contained any longer.
+
+AttackAbility(Ability):
+	# Every attack shall inherit from this.
+	# WIP: how will damage calculation work in oa?
+
+	range : float
+	rate : float
+	damage : float # rename to baseDamage ?
+
+
+HealingAbility(Ability): #WIP
+	# Workers can repair buildings, ships and siege weapons. Buildings heal contained Units, Priests can heal moving units (including siege weapons?)
+	# WIP: priests and workers need an animation for this, buildings not.
+	# Should we write that as an attack with negative
+
+	range : float
+	rate : float
+	damage/reversedamage
+
+ResourceGenerator():
+	# creates constantly resource. Used by relicts.
+
+	resource : Resource
+	rate : float
+
+# Proposal
+UnitTypes(): # Bad name, rename.
+	# This class is used to group several ingame objects. This can be used for
+	# research (e.g. boost all Infantry by +2 speed), AttackAbility (what
+	# can (not) be attacked, CarryAbility and maybe more.
+
+#and probably more. if you know what's missing, add it.
+
+```
+
+Open nyan implementation questions:
+
+4. How do we add new animations?
+
+  -> How would a mod add a new relict-like object that can be carried (the animation)? can we describe composition of frames? How is it done by the original game?
```

Comment:
Original game: Two different units: `monk` and  `monk with relic`

Reply:
I'd do it with the ability to carry things, and a relic is a `Carryable`. To solve the asset combination the thing to be carried could be able to provide a overall carrying graphic, so the `monk-with-relic` graphic is activated as soon as it is picked up.

Reply:
> `ResourceProvider` when garrisoned in the monastry

@TheJJ Now that you mention it, I'd even go further than that and regard the relic as some kind of `BonusItem` or `BonusProvider`. While it's true that AoE2 see relics only as another source for gold income, AoM's relics also provide boni like cheaper unit creation cost, faster research time. RoN has something similar, where rare resources on the map provide a bonus to the player who places a merchant on it (and rare resources can be controlled by multiple people). I think modders would also get very creative with this construct.

Reply:
how would BonusProvider look like?

Reply:
It would be similar to a `ResourceProvider`, but instead of giving a `Resource`, it provides a `Bonus` to the player who controls the `BonusProvider`. Boni could give stats upgrades (like 5% faster unit creation), area of effect buffs (like the generals in RoN), constant flow of resources (e.g. relics and feitoria in AoE2) or give a new ability to a hero who equips the `Mighty Staff of Midgets`, for example. Boni could also be temporary, permanent or tied to an item/unit/building. 

In `nyan` a `Bonus` would probably be utilized as a nyan patch. Similar to the `Tech()` patch. `BonusProvider` on the other hand would be like a `ResourceSpot`:

```
Item():
     name : text

Bonus():
     timeout : float
     updates : set(NyanPatch)

ResourceBonus(Bonus):
     resources : Resource
     rate : float

BonusProvider():
    condition : Condition
    is_temporary : bool
    boni : set(Bonus)

# Content
Relic(Item, BonusProvider):
     name = "relic"
     
     condition = garrisoned
     is_temporary = false
     boni = {RelicBonus}

RelicBonus(ResourceBonus):
     resource = {Gold}
     timeout = -1.0
     rate = 1.0

     GoldBoost<resource>():
         amount += 2.0

     patches = {GoldBoost}
```

Or something like that. I hope i got the nyan syntax right.